### ISSUE TITLE
EHN: `map` more flexible

### ISSUE THREAD

**Issue by pburdet** [hyperspy/hyperspy#496]
In the following example, `map` doesn't work because the function returns data with a different shape.

`map` should support returning any data shape

``` python

from skimage.feature import peak_local_max
import scipy.misc
ims = signals.Signal(scipy.misc.face()).as_image([0,1])
index = im.map(peak_local_max, min_distance=100, num_peaks=4)
for i in range(4):
     m = utils.plot.markers.point(x=index[:, i, 1],
                                  y=index[:, i, 0], color='red')
     ims.add_marker(m)

```

![plot_markers_im](https://cloud.githubusercontent.com/assets/3855190/6921339/2c7bdf44-d7bc-11e4-8a6f-12acb72eb907.png)

**Comment by dnjohnstone** [comment]
I think I faced this issue recently trying to find peaks and extract lattice vectors in images. I dealt with this as in #563 but not sure if a similar solution could be used for the map function - or perhaps I misunderstood the issue.

A related issue is what to do when you essentially have a variable signal range e.g. 4 peaks for the first image, 5 peaks in the second image... Apparently this is discussed in some places as dealing with "ragged arrays" but not sure that there is a standard/good solution.

**Comment by francisco-dlp** [comment]
[dynd](https://github.com/libdynd/libdynd) implements ragged arrays.

**Comment by pburdet** [comment]
Thanks for the answer, my problem here was not ragged array, but when the array returns by ```map``` is a regular array of different shape than the input. In this case, returning the result as a signal would be easy to implement. It would makes ```map``  of a broader use (similarly to the python map)

**Comment by francisco-dlp** [comment]
Once again, this is a nice feature request. However, in order to keep it in for HyperSpy 0.9 someone should be assigned to it (any volunteer?), otherwise we should move it to the wish list.

**Comment by dnjohnstone** [comment]
I'll have a go at adding this functionality ideally in such a way that other analyses within HyperSpy can call the mapping functionality because in trying to find peaks in an image and then convert this to a pair of lattice vectors I have essentially been repeatedly writing such functions. From a brief chat with Francisco we think that the map function should be able to operate in at least 3 modes and below I list the 5 that I think would make it fully general:

Mode 1: output dimension equal to input dimension (currently optimised for speed and memory)
Mode 2: output dimension regular but not equal to input dimension
Mode 3: output dimension ragged and not equal to input dimension
Mode 4: output dimension regular but input dimension ragged
Mode 5: output dimension and input dimension ragged

**Comment by pburdet** [comment]
This is really general. Really good. The three first mode are indeed required for a general ```map```. Mode 2 shouldn't be long to implement.

**Comment by dnjohnstone** [comment]
Having looked into this a bit better, the issue in terms of making it work, is just that 'map' currently overwrites the original data with the output of the function which leads to a broadcasting error if the output does not have the same dimension as the input.  That's easy to resolve but it leads to some questions about how we want 'map' to work in the long run.

Overwriting the data is (as I understand it) implemented to reduce the memory requirements of the method, which makes sense - but in my opinion it's not ideal for what I envisage to be typical use of 'map'... if for example I want to apply some image processing function, it is likely that I'll want to map that function across my data, then have a look at the output, then tweak some parameters and run it again. Such tweaking is going to require an annoying workflow if the data is overwritten each time.

I propose giving 'map' a boolean argument "overwrite" that will be set by default to False but can be set to True if the memory factors are of greatest concern for the user...

Other opinions would be of interest.

**Comment by francisco-dlp** [comment]
Implemented in #1336
